### PR TITLE
Split type-equality-compat package

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

File: type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE CPP                 #-}
+#if __GLASGOW_HASKELL__ <800
+#error "Trying to compiled Data.Type.Equality.Hetero module with GHC <7.6"
```
Comment (RyanGlScott) [inline type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs]: Why not just re-export `(:~~:)(..)` from `Data.Type.Equality` on newer GHCs?

File: type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import Data.Type.Equality
+import Data.Typeable (Typeable)
+
+#if MIN_VERSION_base(4,7,0)
+import Data.Type.Coercion (TestCoercion (..), Coercion (..))
+#endif
+
+-- | Kind heterogeneous propositional equality. Like ':~:', @a :~~: b@ is
+-- inhabited by a terminating value if and only if @a@ is the same type as @b@.
+data (a :: k1) :~~: (b :: k2) where
```
Comment (RyanGlScott) [inline type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs]: This should be `infix 4`. Similarly for `(:~:)(..)`.

File: type-equality-compat/src-old/Data/Type/Equality.hs [hunk 3]
```diff
@@ -28,6 +28,12 @@ module Data.Type.Equality (
     apply,
 #endif
 
+    -- inner and outer not implemented,
```
Comment (RyanGlScott) [inline type-equality-compat/src-old/Data/Type/Equality.hs]: Actually, you _can_ implement `inner` back to 7.0 (and `outer` back to 7.6) using the Leibniz equality trick. See https://gitlab.haskell.org/snippets/1464 for a proof of concept.

File: type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs (hunk 1 above)
Comment (phadej) [reply type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs]: This module does export `:~~:` on newer GHCs? See line 23-25.

This error is a redundancy check to catch "I'm trying to compile this with GHC-7.x"

Comment (RyanGlScott) [reply type-equality-compat/src-hetero/Data/Type/Equality/Hetero.hs]: Oops, you're absolutely right. I think I meant to comment on a different part of the PR.

File: type-equality-compat/type-equality-compat.cabal [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+library
+  default-language: Haskell2010
+  build-depends:    base >=4.3 && <4.14
+
+  if !impl(ghc >=7.8)
+    hs-source-dirs: src-old
+    exposed-modules:  Data.Type.Equality
+
+  if impl(ghc >=8.0 && <8.2)
```
Comment (RyanGlScott) [inline type-equality-compat/type-equality-compat.cabal]: This is what I really meant to comment on. Is there any reason to restrict exposing `Data.Type.Equality.Hetero` to only GHC <8.2? Why not just 8.0 and onwards?

Comment (phadej) [reply type-equality-compat/type-equality-compat.cabal]: Good catch!

**Comment by phadej** [comment]
One more commit. I added you @RyanGlScott as an author, as you contributed a lot :)

**Comment by phadej** [comment]
Ok, let's see what @hesselink replies about `type-equality`. I guess there is no extreme urgency with this package release (i.e. can wait few weeks?)

**Pull request merged**